cpp/include/raft/comms/std_comms.hpp

```diff
@@ -0,0 +1,406 @@
+    ncclResult_t status = call;                                                \
+    ASSERT(ncclSuccess == status, "ERROR: NCCL call='%s'. Reason:%s\n", #call, \
+           ncclGetErrorString(status));                                        \
+  } while (0)
+
+#define NCCL_CHECK_NO_THROW(call)                         \
+  do {                                                    \
+    ncclResult_t status = call;                           \
+    if (status != ncclSuccess) {                          \
+      printf("NCCL call='%s' failed. Reason:%s\n", #call, \
+             ncclGetErrorString(status));                 \
+    }                                                     \
+  } while (0)
+
+namespace raft {
+namespace comms {
+
+size_t getDatatypeSize(const datatype_t datatype) {
+  switch (datatype) {
+    case CHAR:
+      return sizeof(char);
+    case UINT8:
+      return sizeof(uint8_t);
+    case INT:
+      return sizeof(int);
+    case UINT:
+      return sizeof(unsigned int);
+    case INT64:
+      return sizeof(int64_t);
+    case UINT64:
+      return sizeof(uint64_t);
+    case FLOAT:
+      return sizeof(float);
+    case DOUBLE:
+      return sizeof(double);
+  }
+}
+
+ncclDataType_t getNCCLDatatype(const datatype_t datatype) {
+  switch (datatype) {
+    case CHAR:
+      return ncclChar;
+    case UINT8:
+      return ncclUint8;
+    case INT:
+      return ncclInt;
+    case UINT:
+      return ncclUint32;
+    case INT64:
+      return ncclInt64;
+    case UINT64:
+      return ncclUint64;
+    case FLOAT:
+      return ncclFloat;
+    case DOUBLE:
+      return ncclDouble;
+  }
+}
+
+ncclRedOp_t getNCCLOp(const op_t op) {
+  switch (op) {
+    case SUM:
+      return ncclSum;
+    case PROD:
+      return ncclProd;
+    case MIN:
+      return ncclMin;
+    case MAX:
+      return ncclMax;
+  }
+}
+
+class std_comms : public comms_iface {
+ public:
+  std_comms() = delete;
+
+  /**
+   * @brief Constructor for collective + point-to-point operation.
+   * @param comm initialized nccl comm
+   * @param ucp_worker initialized ucp_worker instance
+   * @param eps shared pointer to array of ucp endpoints
+   * @param size size of the cluster
+   * @param rank rank of the current worker
+   */
+  std_comms(ncclComm_t comm, ucp_worker_h ucp_worker,
+            std::shared_ptr<ucp_ep_h *> eps, int size, int rank)
+    : _nccl_comm(comm),
+      _ucp_worker(ucp_worker),
+      _ucp_eps(eps),
+      _size(size),
+      _rank(rank),
+      _next_request_id(0) {
+    initialize();
+    p2p_enabled = true;
+  };
+
+  /**
+   * @brief constructor for collective-only operation
+   * @param comm initilized nccl communicator
+   * @param size size of the cluster
+   * @param rank rank of the current worker
+   */
+  std_comms(ncclComm_t comm, int size, int rank)
+    : _nccl_comm(comm), _size(size), _rank(rank) {
+    initialize();
+  };
+
+  virtual ~std_comms() {
```

Comment:
Originally this comms was a separate library in cuml, which didn’t have the device buffer available as a dependency (it would have created a circular dependency) so we just used cudamalloc. It’s only allocating a single byte for use in `barrier` but I do think device buffer would be cleaner.